lib/components/app/session-timeout.tsx

```diff
@@ -0,0 +1,124 @@
+/* eslint-disable @typescript-eslint/ban-ts-comment */
+/* eslint-disable react/prop-types */
+import { Button, Modal } from 'react-bootstrap'
+import { connect } from 'react-redux'
+import { FormattedMessage } from 'react-intl'
+import React, { Component } from 'react'
+
+import * as uiActions from '../../actions/ui'
+
+/**
+ * This component makes the current session timeout
+ * by displaying a timeout warning one minute before the timeout,
+ * and by reloading the initial URL if there is no user-initiated
+ * actions within the timeout window.
+ */
+class SessionTimeout extends Component {
+  state = {
+    showTimeoutWarning: false,
+    timeoutObject: null,
+    timeoutStartMillis: 0
+  }
+
+  componentDidMount() {
+    // Wait one second or so after loading before probing for changes
+    // so that initialization actions can complete.
+    setTimeout(this.handleAfterInitialActions, 1500)
+  }
+
+  componentWillUnmount() {
+    // @ts-ignore SessionTimeout is not typed yet
+    clearInterval(this.state.timeoutObject)
+  }
+
+  handleAfterInitialActions = () => {
+    this.setState({
+      timeoutObject: setInterval(this.handleTimeoutWatch, 10000),
+      timeoutStartMillis: new Date().valueOf()
+    })
+  }
+
+  handleTimeoutWatch = () => {
+    // @ts-ignore SessionTimeout is not typed yet
+    const { lastActionMillis, sessionTimeoutSeconds, startOverFromInitialUrl } =
+      this.props
+    if (lastActionMillis > this.state.timeoutStartMillis) {
```

Comment:
Is the purpose of this so that the timer doesn't start until the user makes their first action on the page? Can you add a comment about that? Took a bit of thinking on my part to finally figure that out

Reply:
With this check, actions that happened before the session timer is started are simply ignored for timeout purposes. I'll add a comment about that.